File: LeetCode/22/main.go
func gen(s string, open int, close int, n int, ans *[]string) {

    if len(s) == 2 * n {
        *ans = append(*ans, s)
        return
    }

    if open < n {
        gen(s + "(", open + 1, close, n, ans)
    }

    if close < open {
        gen(s + ")", open, close + 1, n, ans)
    }

}

func generateParenthesis(n int) []string {
    ans := []string{}
    gen("", 0, 0, n, &ans)
    return ans
}

File: LeetCode/22/main.cpp
class Solution {
public:
  void gen(string s, int n, int open, int close, vector<string> &ans) {

    if (s.length() == 2 * n) {
      ans.emplace_back(s);
      return;
    }

    if (open < n) {
      // Add (
      gen(s + "(", n, open + 1, close, ans);
    }

    if (close < open) {
      // Add )
      gen(s + ")", n, open, close + 1, ans);
    }

    return;
  }

  vector<string> generateParenthesis(int n) {
    vector<string> ans;
    string s = "";

    gen(s, n, 0, 0, ans);

    return ans;
  }
};
